File: ProvidedClients/JuliaClient/GameState.jl
using .PlayerInterfaceClients
using Random: bitrand

import .PlayerInterfaceClients: GameUpdateMessage, UpdatedCell, PlayerScore

include("Snake.jl")

mutable struct Cell
    Address::Vector{Int};
    HasFood::Bool;
    HasPlayer::Bool;

    function Cell(address::Vector{Int}, food::Bool, player::Bool)
        return new(address, food, player);
    end
end

mutable struct GameState
    Cells::Array{Cell};
    Dimensions::NTuple{N,Int} where N
    Snakes::Array{Snake}
    function GameState(dims::NTuple{N,Int}, startAddress::Vector{Int}, playerName::String) where N
        snakes = Snake[];
        push!(snakes, Snake(Int.(startAddress) .+ 1, playerName));
        cells = Array{Cell}(undef, dims);
        return new(cells, dims, snakes);
    end
end

function GetCell(state::GameState, address::Vector{Int})
    if !isassigned(state.Cells, address...)
        state.Cells[address...] = Cell(address, false, false);
    end 
    return state.Cells[address...];
end

function SetCell(state::GameState, address::Vector{Int}, food::Bool, player::Bool)
    if !isassigned(state.Cells, address...)
        state.Cells[address...] = Cell(address, food, player);
    else
        cell = state.Cells[address...];
        cell.HasFood = food;
        cell.HasPlayer = player;
    end 
end

function IsValidAndEmpty(state::GameState, address::Vector{Int})
    #println(address);
    if !checkbounds(Bool, state.Cells, address...)
        return false
    end
    return !(GetCell(state, address).HasPlayer);
end

function HasFood(state::GameState, address::Vector{Int})
    if !checkbounds(Bool, state.Cells, address...)
        return false
    end
    return (GetCell(state, address).HasFood);
end

function GetValidAddress(state::GameState, address::Vector{Int})
    while true
        result = copy(address);
        dim = rand(1:length(state.Dimensions),1)[1];
        direction = bitrand(1)
        if direction[1]
            result[dim] += 1;
        else
            result[dim] -= 1;
        end
        if IsValidAndEmpty(state, result)
            return result;
        end
    end
end

function GetMoves(gameState::GameState)
    moves = Move[];
    for snake in gameState.Snakes
        println("making move for snake " * snake.Name);
        address = GetValidAddress(gameState, snake.Head);
        snake.Head = address;
        push!(snake.Segments, address);
        if HasFood(gameState, address)
            snake.Length = snake.Length + 1;
        else
            popfirst!(snake.Segments);
        end
        move = Move(playerIdentifier=playerIdentifier, snakeName=snake.Name, nextLocation=Int32.(address .- 1));
        push!(moves, move);
    end
    return moves;
end

function GetSplits(gameState::GameState)
    splits = SplitRequest[];
    for snake in gameState.Snakes
        if snake.Length > 2 && length(gameState.Snakes) < 11
            println("old snake:");
            for seg in snake.Segments
                println(seg);
            end
            snake.Length = snake.Length - 1;
            snake.KidCount += 1;
            newHead = popfirst!(snake.Segments); #this also removes it from snake
            println(newHead);
            newSnake = Snake(newHead, snake.Name * "." * string(snake.KidCount));
            push!(newSnake.Segments, newHead);
            push!(gameState.Snakes, newSnake);
            address = GetValidAddress(gameState, newSnake.Head);
            println(address);
            newSnake.Head = address;
            push!(newSnake.Segments, address);
            if HasFood(gameState, address)
                newSnake.Length = 2;
            else
                popfirst!(newSnake.Segments);
            end
            split = SplitRequest(playerIdentifier=playerIdentifier, oldSnakeName=snake.Name, newSnakeName=newSnake.Name, snakeSegment=1, nextLocation=Int32.(address .- 1));
            push!(splits, split);
        end
    end
    return splits;
end

function UpdateGameState(state::GameState, update::GameUpdateMessage)
    for cell in update.updatedCells
        address = Int.(cell.address) .+ 1;
        SetCell(state, address, cell.foodValue > 0, !isempty(cell.player));
    end
end

File: ProvidedClients/JavascriptClient/GameState.js
const { Cell } = require("./Cell");
const { Snake } = require("./Snake");

class GameState {
    cells = [];
    dimensions = [];
    snakes = [];
    playerIdentifier = "";

    constructor(dims, startAddress, playerName, playerIdentifier) {
        this.snakes.push(new Snake(startAddress, playerName));
        this.dimensions = dims;
        this.cells = this.createNDimArray(dims);
        this.playerIdentifier = playerIdentifier;
    }

    createNDimArray(dimensions) {
        if (dimensions.length > 0) {
            var dim = dimensions[0];
            var rest = dimensions.slice(1);
            var newArray = new Array();
            for (var i = 0; i < dim; i++) {
                newArray[i] = this.createNDimArray(rest);
            }
            return newArray;
        } else {
            return undefined;
        }
    }

    checkBounds(address) {
        for(var i = 0; i<address.length;i++){
            if(address[i] < 0){
                return false;
            }
            if(address[i] >= this.dimensions[i]){
                return false;
            }
        }
        return true;
    }

    getCell(address) {
        var cell = this.cells[address];
        if (!cell) {
            cell = new Cell(address, false, false);
        }
        this.cells[address] = cell;
        return cell;
    }

    update(gameUpdate) {
        for(var i = 0; i < gameUpdate.updatedCells.length; i++){
            var cell = this.getCell(gameUpdate.updatedCells[i].address);
            cell.hasFood = gameUpdate.updatedCells[i].foodValue > 0;
            cell.hasPlayer = gameUpdate.updatedCells[i].player.length > 0;
            this.cells[gameUpdate.updatedCells[i].address] = cell;
        }
    }

    getNextAddress(address) {
        while (true) {
            var newaddr = [...address];
            var dim = Math.floor(Math.random() * this.dimensions.length);
            var dir = Math.floor(Math.random() * 2);
            if (dir > 0) {
                newaddr[dim] += 1;
            } else {
                newaddr[dim] -= 1;
            }
            if (this.checkBounds(newaddr)) {
                var cell = this.getCell(newaddr);
                if (cell.hasPlayer == false) {
                    return newaddr;
                }
            }
        }
    }

    getMoves() {
        var moves = [];
        for (var i = 0; i < this.snakes.length; i++) {
            var snake = this.snakes[i];
            var nextLocation = this.getNextAddress(snake.head);
            snake.head = nextLocation;
            var cell = this.getCell(nextLocation);
            snake.segments.push(nextLocation);

            if (cell.hasFood) {
                snake.length += 1;
            } else {
                snake.segments = snake.segments.slice(1);
            }
            moves.push({
                playerIdentifier: this.playerIdentifier,
                snakeName: snake.name,
                nextLocation: nextLocation
            });
        }

        return moves;
    }

    getSplits() {
        var splits = [];
        for (var i = 0; i < this.snakes.length; i++) {
            var snake = this.snakes[i];
            if(snake.length > 2 && this.snakes.length < 11) {
            
                snake.length -= 1;
                snake.kidcount += 1;
                var newHead = snake.segments[0];
                snake.segments = snake.segments.slice(1);
                var newSnake = new Snake(newHead, snake.name + "." + snake.kidcount) ;
                this.snakes.push(newSnake);
                var address = this.getNextAddress(newHead)
                newSnake.head = address
                var cell = this.getCell(address);
                newSnake.segments.push(address);
                if (cell.hasFood){
                    newSnake.length += 1
                } else {
                    newSnake.segments.shift();
                }

                splits.push({
                    playerIdentifier: this.playerIdentifier,
                    newSnakeName: newSnake.name,
                    oldSnakeName: snake.name,
                    snakeSegment: 1,
                    nextLocation: address
                });
            }
        }

        return splits;
    }
}
exports.GameState = GameState;
